### ISSUE TITLE
Computing result type with type helpers for views.

### ISSUE THREAD

**Issue by tower120** [ericniebler/range-v3#932]
Need some way to **PRE**compute result view type. 

Having concrete (not auto) type as function return, allow to move implementation from header to source file (this is important for breaking circular dependency). 
Example:

```cpp
// CarTypes.h
class CarTypes {
protected:
    using m_wheels_t = std::vector<std::unique_ptr<Wheel>>;
public:
    // this not work!
    // using wheels_t = ranges::indirect_view< m_wheels_t>;  

    // here, need helper ranges::indirect_view_t, instead of this:
    using wheels_t = ranges::indirect_view< ranges::view::all_t<m_wheels_t > >;
}
```

```cpp
// Car.h
class Car : public CarTypes {
public:
    wheels_t wheels() {  // return concrete type, not auto. Now it is possible to move wheels() to .cpp file
	return m_wheels | ranges::view::indirect;     // "return {m_wheels};"  this would be even nicer
    }  
private:
  m_wheels_t m_wheels;
};
```

```cpp
// CarsHelperTypes.h
// imagine that we can't include Car.h here, due to cyclic dependency
class Car;

// separating all "interface" types to separate class and header can make life sometimes easier.
// This allows to work with types (#include "xxxxxTypes.h") without need of including implementation.
struct CarsHelperTypes{
       const auto fn = [](auto* car){
          return yield_from(car->wheels());
      };
     using cars_t = std::vector<Car*>;
     using all_wheels_t = for_each_view_t<cars_t , decltype(fn)>;
};

// CarsHelper.h
struct CarsHelper : CarsHelperTypes{
   cars_t cars;
   all_wheels_t all_wheels();
}
```

```cpp
// CarsHelper.cpp
#include " CarsHelper.h"

// including implementation now!
#include "Car.h"

CarsHelperTypes::all_wheels_t CarsHelper::all_wheels(){
      return view::for_each(cars, fn);
}
```

It is not always possible to separate return types from implementation like this (thou I think often enough to have practical use), but when it is - we can avoid slowness of  `any_view`.

So, to summarize:
 * It would be good, if each view function return exactly corresponding type; like: `ranges::view::indirect -> ranges::indirect_view`  (maybe do decaying/other type conversion directly in indirect_view? Utilize c++17 class template argument deduction to get rid of functions-makers at all? *Or whatever reason for these helper functions to exists - pipeability?*)
 * Either have some helper types like `inidrect_t = ranges::indirect_view< ranges::view::all_t<m_wheels_t > >`. If you go this way, solutions with decltype will not work, because declval requires complete type; e.g. something like this will not work in long term:
```cpp
   template<class Rng, class Fun>
   using transform_t = decltype(ranges::view::transform(std::declval<Rng&>(), std::declval<Fun>()));
```
With this you can't just forward declare `Fun` class, it need to be included/known at the very place where you spell `transform_t`.

P.S. *It is, of course possible to have "homebrew" helper library for cases like this. But it would be good to have some canonical way to compute return type.* 

N.B. At first, I thought this kind of problems will fly away with modules for good... Just type auto and return range in header. But it seems, that module system will not solve class circular dependency - breaking definition and implementation still is a required thing with "proclaimed ownership declaration" (see [this article](https://www.codeproject.com/Articles/1214398/Modules-for-Modern-Cplusplus)(last section),  and [at least n4681](http://open-std.org/JTC1/SC22/WG21/docs/papers/2017/n4681.pdf) )

**Comment by tower120** [comment]
I think I have good solution for this.
```cpp
compose_view<class Src, class Transformation, class ...Transformations> 
    = Transformation< compose_view<Src, Transformations...> >;

compose_view<class Src, class Transformation> 
    = Transformation< view::all_t<Src> >;
```
Usage:
```cpp
using List = std::vector<int>;
List list;

using Get = compose_view<List, indirect_view, const_view >;
Get get();

// in .cpp
Get get(){
    return list | view::indirect | view::const_;
}
```
Multi-parameter views:
```cpp
struct GetFn{
   auto& operator()(const Data& data) const{ return data.i; }
};
using Get = compose_view<List, compose_bind<transform_view, GetFn>>
Get get();

// in .cpp
Get get(){
    return list | view::transform(GetFn{});
}
```

----

Also, we could make `compose_view` view_adaptor, with two constructors - one accept view chain (as previously), another - source range as argument:

```cpp
Get get(){
    return {list};
}
```

Second option will work, only if all adapting views have unary constructors. But that's quite often!
For example, with small change in `view::transform`, the following will work:
```cpp
struct GetFn{
   auto& operator()(const Data& data) const{ return data.i; }
};
using Get = compose_view<List, compose_bind<transform_view, GetFn> >;

// in .cpp
Get get(){
    return {list};
}
```

**Comment by CaseyCarter** [comment]
For the first example, `wheels_t` is easily computed as `decltype(std::declval<m_wheels_t&>() | ranges::view::indirect)` ([Fully worked example](https://wandbox.org/permlink/ciKGVoSoKmyPxHyV)).

In the second example, `all_wheels_t` isn't computable by any mechanism. Even simple properties like the value type and reference type of the resulting range depend on how `Car` is completed. Range-v3 certainly can't implement an oracle that divines whether `car->wheels()` returns a reference to an array of 3 integers or returns a `std::map<std::string, std::vector<double>>` by value.

I'm not sure there's a problem here that the library can solve more effectively than a user can themselves via application of `decltype`.

**Comment by tower120** [comment]
> For the first example, wheels_t is easily computed as decltype(std::declval<m_wheels_t&>() | ranges::view::indirect)

Simplicity of such computation depends on the amount of `std::declval`'s in your chain. `decltype` computation with large amount of std::declval 's quickly becomes quite verbose (non-unary constructible views):

```cpp
using Get = decltype(
    std::declval<m_wheels_t&>() 
    | view::transform(std::declval<Fn>())
    | my_filter_view(std::declval<FilterType1>(), std::declval<FilterType2>())
);
```
Not saying, that you must repeat that chain at implementation site.

> Range-v3 certainly can't implement an oracle that divines whether car->wheels() returns a reference to an array of 3 integers or returns a std::map<std::string, std::vector<double>> by value.

Ok. Let's change `CarsHelperTypes` a  bit.

```cpp
// CarsHelperTypes.h
#include "CarTypes.h"   // for wheels_t 
class Car;

struct CarsHelperTypes{
     // or replace this with functor
     static m_wheels_t get_wheels(Car*);  // -> std::vector<Wheel&>

     const auto fn = [](auto* car){
         return yield_from(get_wheels);
     };

     using cars_t = std::vector<Car*>;
     using all_wheels_t = for_each_view_t<cars_t , decltype(fn)>;
};

// CarsHelperTypes.cpp
#include "CarsHelperTypes.h"
#include "Car.h"
CarsHelperTypes::m_wheels_t CarsHelper::get_wheels(Car* car){
      return car->wheels();
}
```
Now (I simplified a bit):
```
m_wheels_t = std::vector<Wheel*>;
wheels_t   = ranges::indirect_view<m_wheels_t> = std::vector<Wheel&>;
all_wheels_t depends on cars_t  and wheels_t.
```
`all_wheels_t` computes without any knowledge about `Car` internals.


> I'm not sure there's a problem here that the library can solve more effectively than a user can themselves via application of decltype.

Library can provide mechanism for simplifying construction of such objects. Now, with `decltype`, you literally have to write view chain twice:
```cpp
using wheels_t = decltype(std::declval<m_wheels_t&>() | ranges::view::indirect | ranges::view::const_);
wheels_t wheel() const {
     return m_wheels | ranges::view::indirect | ranges::view::const_;
}
```
This particular chain can be simplified:
```cpp
using wheels_t = compose_view<m_wheels_t, ranges::view::indirect, ranges::view::const_>;
wheels_t wheel() const {
     return wheels_t{m_wheels};
}
```
Non-unary views can be covered too https://github.com/ericniebler/range-v3/pull/1172#issue-274866357 .

**Comment by ericniebler** [comment]
I agree with Casey; I don't think this is a problem that needs to be solved.

**Comment by tower120** [comment]
Because result type CAN be computed with `decltype`. And the cases where you actually need to separate function definition from implementation is rare enough? (despite the fact of code repetition)

IOW - use cases are rare and exotic?
